src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1577,6 +1595,66 @@ public override BoundNode VisitCall(BoundCall node)
             return null;
         }
 
+        private bool MethodReturnsTrueWhenNotNull(MethodSymbol method)
+        {
+            if (method.ReturnType.SpecialType != SpecialType.System_Boolean)
+            {
+                return false;
+            }
+
+            return method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_String_IsNullOrEmpty));
+        }
+
+        private void VisitTrueWhenNotNull(BoundExpression operand, TypeSymbol boolType)
+        {
+            if (!operand.Type.IsReferenceType)
+            {
+                return;
+            }
+
+            // operand == null || someUnknownBoolean
+            BoundExpression equivalent = LogicalOr(
+                new BoundBinaryOperator(operand.Syntax, BinaryOperatorKind.Equal, operand,
+                    new BoundLiteral(operand.Syntax, ConstantValue.Null, operand.Type),
+                    constantValueOpt: null, methodOpt: null, LookupResultKind.Viable, boolType),
+                new BoundLiteral(operand.Syntax, constantValueOpt: null, boolType));
+
+            // We only need to update the null-state, so we'll do the flow analysis without diagnostics
+            bool savedDisableDiagnostics = _disableDiagnostics;
+            _disableDiagnostics = true;
+
+            Visit(equivalent);
+
+            _disableDiagnostics = savedDisableDiagnostics;
+        }
+
+        internal static BoundBinaryOperator LogicalOr(BoundExpression left, BoundExpression right)
+        {
+            return new BoundBinaryOperator(left.Syntax, BinaryOperatorKind.LogicalBoolOr, left, right, constantValueOpt: null, methodOpt: null, LookupResultKind.Viable, left.Type);
+        }
+
+        private bool MethodEnsuresTrueWhenExits(MethodSymbol method)
+        {
+            return method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert1)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert2)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert3)) ||
+                method.Equals(compilation.GetWellKnownTypeMember(WellKnownMember.System_Diagnostics_Debug_Assert4));
+        }
+
+        private void VisitTrueWhenExits(BoundExpression condition)
+        {
+            // We only need to update the null-state, so we'll do the flow analysis without diagnostics
+
+            bool savedDisableDiagnostics = _disableDiagnostics;
+            _disableDiagnostics = true;
+
+            // if (!condition) throw;
+            VisitIfStatement(new BoundIfStatement(condition.Syntax, SyntheticBoundNodeFactory.NotCore(condition),
+                new BoundThrowStatement(condition.Syntax, expressionOpt: null), alternativeOpt: null));
```

Comment:
Can we implement this visit method without synthesizing nodes, by mimicking the behavior of `VisitIfStatement` and `VisitThrowExpression`?

Reply:
I didn't want to duplicate that logic, but we can discuss some alternatives.